#  Initialization



 ## Import Libraries

In [8]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from matplotlib.ticker import StrMethodFormatter
%matplotlib inline
import scipy.stats as stats
import seaborn as sns
from imblearn.over_sampling import SMOTE
from imblearn.over_sampling import RandomOverSampler
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn import metrics
from sklearn.metrics import confusion_matrix
from sklearn.metrics import roc_curve, auc
from sklearn.metrics import classification_report
from sklearn.neighbors import KNeighborsClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.ensemble import GradientBoostingClassifier
from imblearn.over_sampling import SMOTE
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score, roc_auc_score, precision_recall_fscore_support, roc_curve, f1_score
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score, roc_auc_score, precision_recall_fscore_support, roc_curve, f1_score
from sklearn.metrics import roc_auc_score 
from sklearn.model_selection import cross_validate


## Loading Data



In [9]:
#To run the code, you can change this path to the desired one
df=pd.read_json(r"/Users/..../Desktop/transactions.txt",lines=True)

In [10]:
#Substitute empty values with Nan
df=df.replace('',np.NaN)

In [11]:
df.shape

(786363, 29)


We can see our data has 786,363 instances and each one of them has 29 features

In [12]:
#df.head()


Below we can see some basic statistics from our data set.

In [13]:
df[['creditLimit','availableMoney','transactionAmount','currentBalance']].describe()

creditLimit  availableMoney  transactionAmount  currentBalance
count  786363.000000   786363.000000      786363.000000   786363.000000
mean    10759.464459     6250.725369         136.985791     4508.739089
std     11636.174890     8880.783989         147.725569     6457.442068
min       250.000000    -1005.630000           0.000000        0.000000
25%      5000.000000     1077.420000          33.650000      689.910000
50%      7500.000000     3184.860000          87.900000     2451.760000
75%     15000.000000     7500.000000         191.480000     5291.095000
max     50000.000000    50000.000000        2011.540000    47498.810000


The number of unique values in each column are shown below

In [14]:
df.nunique(axis=0)

accountNumber                 5000
customerId                    5000
creditLimit                     10
availableMoney              521915
transactionDateTime         776637
transactionAmount            66038
merchantName                  2490
acqCountry                       4
merchantCountryCode              4
posEntryMode                     5
posConditionCode                 3
merchantCategoryCode            19
currentExpDate                 165
accountOpenDate               1820
dateOfLastAddressChange       2184
cardCVV                        899
enteredCVV                     976
cardLast4Digits               5245
transactionType                  3
echoBuffer                       0
currentBalance              487318
merchantCity                     0
merchantState                    0
merchantZip                      0
cardPresent                      2
posOnPremises                    0
recurringAuthInd                 0
expirationDateKeyInMatch         2
isFraud             


Now we can see the percentage of Null values in each columns

In [15]:
round(df.isnull().sum()*100/len(df),2)


accountNumber                 0.00
customerId                    0.00
creditLimit                   0.00
availableMoney                0.00
transactionDateTime           0.00
transactionAmount             0.00
merchantName                  0.00
acqCountry                    0.58
merchantCountryCode           0.09
posEntryMode                  0.52
posConditionCode              0.05
merchantCategoryCode          0.00
currentExpDate                0.00
accountOpenDate               0.00
dateOfLastAddressChange       0.00
cardCVV                       0.00
enteredCVV                    0.00
cardLast4Digits               0.00
transactionType               0.09
echoBuffer                  100.00
currentBalance                0.00
merchantCity                100.00
merchantState               100.00
merchantZip                 100.00
cardPresent                   0.00
posOnPremises               100.00
recurringAuthInd            100.00
expirationDateKeyInMatch      0.00
isFraud             


There are 6 columns that all of their values are Null, so I remove these columns.

In [16]:
df.drop(['echoBuffer','merchantCity','merchantState','merchantZip','posOnPremises','recurringAuthInd'], axis=1, inplace=True)
df=df.reset_index(drop=True)
